third_party/move/move-compiler-v2/src/pipeline/split_critical_edges_processor.rs

```diff
@@ -0,0 +1,193 @@
+    fn transform(&mut self) {
+        let bytecodes = std::mem::take(&mut self.data.code);
+        for bytecode in bytecodes {
+            self.transform_bytecode(bytecode)
+        }
+    }
+
+    /// Transforms a bytecode
+    fn transform_bytecode(&mut self, bytecode: Bytecode) {
+        match bytecode {
+            Bytecode::Branch(attr_id, l0, l1, t) => self.transform_branch(attr_id, l0, l1, t),
+            // Edge of a `Jump` can never be critical since the source node only have one out edge
+            _ => self.emit(bytecode),
+        }
+    }
+
+    /// Transforms a branch instruction by splitting the critical out edges
+    pub fn transform_branch(&mut self, attr_id: AttrId, l0: Label, l1: Label, t: TempIndex) {
+        match (
+            self.split_critical_edge(attr_id, l0),
+            self.split_critical_edge(attr_id, l1),
+        ) {
+            (None, None) => self.emit(Bytecode::Branch(attr_id, l0, l1, t)),
+            (None, Some((l1_new, code))) => {
+                self.emit(Bytecode::Branch(attr_id, l0, l1_new, t));
+                self.emit_codes(code)
+            },
+            (Some((l0_new, code)), None) => {
+                self.emit(Bytecode::Branch(attr_id, l0_new, l1, t));
+                self.emit_codes(code)
+            },
+            (Some((l0_new, code0)), Some((l1_new, code1))) => {
+                self.emit(Bytecode::Branch(attr_id, l0_new, l1_new, t));
+                self.emit_codes(code0);
+                self.emit_codes(code1);
+            },
+        }
+    }
+
+    /// `label` is the target of some branch instruction
+    /// If label has multiple sources, returns
+    /// - a fresh label
+    /// - a new empty block that
+    ///     - starts with the fresh label
+    ///     - jumps to `label`
+    /// otherwise returns `None`
+    fn split_critical_edge(
+        &mut self,
+        attr_id: AttrId,
+        label: Label,
+    ) -> Option<(Label, Vec<Bytecode>)> {
+        // label is in `srcs_count` by construction
+        if *self.srcs_count.get(&label).expect("srcs count") > 1 {
+            Some(self.split_edge(attr_id, label))
+        } else {
+            None
+        }
+    }
+
+    /// Returns
+    /// - a fresh label
+    /// - a new empty block that
+    ///     - starts with the fresh label
+    ///     - jumps to `label`
+    fn split_edge(&mut self, attr_id: AttrId, label: Label) -> (Label, Vec<Bytecode>) {
+        let new_label = self.gen_fresh_label();
+        let code = vec![
+            Bytecode::Label(attr_id, new_label),
+            Bytecode::Jump(attr_id, label),
+        ];
+        (new_label, code)
+    }
+
+    /// Generates a fresh label
+    fn gen_fresh_label(&mut self) -> Label {
+        let new_label = Label::new(
+            if self.labels.is_empty() {
+                0
+            } else {
+                self.labels.iter().next_back().expect("label").as_usize() + 1
```

Comment:
There is an implicit assumption here that all labels in the set have values < size of set. We should state this assumption here implicitly, or better yet, not depend on any such assumption (eg., keep track of max and allocate based on it).

Reply:
Since we are using `BTreeSet`, `self.labels.iter().next_back()` should return the largest label (label is wrapped u16)?

Reply:
Ah, that's true. Disregard my comment.

Reply:
It would make the code clearer if you use a local temporary:
```
let max_label = self.labels.iter().next_back().expect("label");
max_label.as_usize() +1
```

Reply:
done